Implement dynamic provisioner for AWS EBS

jsafrane opened the pull request:
**Do not merge until PR #14537 is merged!**

Depending on PR #14537, this is implementation of a dynamic provisioner for Amazon. It does not implement any QoS classes, it creates volumes with the default setting (=cheapest/slowest?).

Also, it attaches these tags to the volumes so the system admin can see what created the volume and for which PVC:
"kubernetes.io/created-for/PVC/Namespace"
"kubernetes.io/created-for/PVC/Name"

The first commit is cherry-picked from aforementioned PR and I'll remove it when the PR is merged.

markturansky commented:
LGTM.

@justinsb mind reviewing?  Ignore the controller commit.

markturansky commented:
needs rebase.

can you squash the last two?

jsafrane commented:
Squashed and rebased

justinsb commented:
This LGTM.  I do think that we should probably create gp2 volumes by default; as they are not much more expensive and are generally much faster: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html

I'm not sure why I didn't do this in the first place - I guess because CreateVolume was originally just for tests and we were creating small sizes where arguably magnetic is the better deal.

But we can easily fix this later (and probably add support for different QOS classes)

liggitt commented on pkg/controller/persistentvolume/types.go:
```diff
@@ -26,10 +26,25 @@ import (
+	cloudVolumeCreatedForNamespaceTag = "kubernetes.io/created-for/PVC/Namespace"
+	// Name of a tag attached to a real volume in cloud (e.g. AWS EBS or GCE PD)
+	// with name of a persistent volume claim used to create this volume.
+	cloudVolumeCreatedForNameTag = "kubernetes.io/created-for/PVC/Name"
```
lowercase these?

markturansky replied:
We can also tag with the pv name as well as the claim name:

```
kubernetes.io/created-for/pv-name
kubernetes.io/created-for/pvc-namespace
kubernetes.io/created-for/pvc-name
```

This ties the EBS volume (or whichever) to a PV, a namespace, and a PVClaim.

jsafrane replied:
swithed to lowercase, `kubernetes.io/created-for/pvc/namespace` and `kubernetes.io/created-for/pvc/name`

markturansky commented on pkg/volume/aws_ebs/aws_ebs.go:
```diff
@@ -348,3 +389,67 @@ func (c *awsElasticBlockStoreCleaner) TearDownAt(dir string) error {
+		ObjectMeta: api.ObjectMeta{
+			GenerateName: "pv-aws-",
+			Labels: map[string]string{
+				"createdby": "aws_ebs_dynamic_provisioner",
```
Needs "kubernetes.io/" prefix.  Same for all provisioners.

We're not to use naked labels on an object.  We leave that empty namespace for the user.

jsafrane replied:
Done, also moved from `Labels` to `Annotations`, following HostPath example.

markturansky commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1216,15 +1217,15 @@ func (aws *AWSCloud) DetachDisk(instanceName string, diskName string) error {
 	volSize := (int64(volumeOptions.CapacityMB) + 1023) / 1024
 	request.Size = &volSize
-	response, err := aws.ec2.CreateVolume(request)
+	response, err := s.ec2.CreateVolume(request)
```
What EBS VolumeType is this by default?

We have a request for "gp2 SSD".

jsafrane replied:
AWS defaults to 'standard', i.e. Magnetic.

markturansky replied:
Can you add some bits to make the default for the provisioner ssd?

jsafrane replied:
Done, in an extra commit.

jsafrane commented:
Rebased to latest version of #14537 and tested on AWS.

jsafrane commented:
@k8s-bot test this

markturansky committed:
Added PersistentVolumeController

jsafrane committed:
Implement Creater and Deleter interfaces for AWS EBS.

Also mark the created EBS volumes with tags, so the admin knows
who/what created the volumes.

jsafrane committed:
Use SSD as default volume type.

General purpose SSD ('gp2') volume type is just slighly more expensive than
Magnetic ('standard' / default in AWS), while the performance gain is pretty
significant.

So far, the volumes were created only during testing, where the extra cost
won't make any difference. In future, we plan to introduce QoS classes, where
users could choose SSD/Magnetic depending on their use cases.

'gp2' is just the default volume type for (hopefuly) short period before these
QoS classes are implemented.

markturansky commented:
@saad-ali this next, please?

We'll need to rebase this after #14537 is merged.

saad-ali commented:
Taking a look. Will try to tease apart the changes from 14537

saad-ali commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -107,3 +108,42 @@ func (util *AWSDiskUtil) DetachDisk(c *awsElasticBlockStoreCleaner) error {
 }
+
+func (util *AWSDiskUtil) DeleteVolume(d *awsElasticBlockStoreDeleter) error {
+	cloud := d.plugin.host.GetCloudProvider()
```
nit: this code to fetch and cast the cloud provider gets repeated a lot, factor it out into a util method, if possible.

saad-ali commented on pkg/volume/aws_ebs/aws_util.go:
```diff
@@ -107,3 +108,42 @@ func (util *AWSDiskUtil) DetachDisk(c *awsElasticBlockStoreCleaner) error {
+	}
+
+	// convert to MB with rounding
+	volSize := int((c.options.Capacity.Value() + mega - 1) / mega)
```
Could you add a comment on how this work.

saad-ali commented on pkg/volume/aws_ebs/aws_ebs.go:
```diff
@@ -349,3 +390,67 @@ func (c *awsElasticBlockStoreCleaner) TearDownAt(dir string) error {
+	}
+	pv.Spec.PersistentVolumeSource.AWSElasticBlockStore.VolumeID = volumeID
+	pv.Spec.Capacity = api.ResourceList{
+		api.ResourceName(api.ResourceStorage): resource.MustParse(fmt.Sprintf("%d", size)),
```
No units suffix? Are they assumed?

saad-ali commented:
Overall looks good, some minor comments (and needs rebase).

saad-ali commented:
Ignoring this PR, since it appears to be replaced by #18601

jsafrane commented:
Closing in favor of https://github.com/kubernetes/kubernetes/pull/18601
